File: cpp-examples/classes/advanced/Book.cpp
#include "Book.h"
#include <iostream>

using std::string;
using std::cout;
using std::endl;
using std::to_string;

int Book::bookCount = 0; //Initialize Static member

// Constructor with all parameters
Book::Book(const string& title, const string& author, int year) : title(title), author(author), year(year) {
    bookCount++;
}

// Constructor overload with default year
Book::Book(const string& title, const string& author) : Book(title, author, 0) {}

// Destructor
Book::~Book() {
    cout << "Destroying " << title << endl;
    bookCount--;
}

// Print details about the book
void Book::printDetails() const {
    cout << "Title: " << title
              << ", Author: " << author
              << ", Year: " << (year ? to_string(year) : "Unknown") << endl;
}

// Update book title
void Book::updateTitle(const string& newTitle) {
    this->title = newTitle; // 'this' is used to refer to the current object instance
}

// Static function to get the number of books
int Book::getBookCount() {
    return bookCount;
}

File: python-examples/classes/advanced/Book.py
class Book:
    book_count = 0  # Class variable to keep track of the number of books

    def __init__(self, title, author, year=None):
        self.title = title
        self.author = author
        self.year = year
        Book.book_count += 1

    def __del__(self):
        print(f"Deleting {self.title}")
        Book.book_count -= 1

    def print_details(self):
        year_display = self.year if self.year else "Unknown"
        print(f"Title: {self.title}, Author: {self.author}, Year: {year_display}")

    def update_title(self, new_title):
        self.title = new_title  # 'self' refers to the instance the method is called on

    @classmethod
    def get_book_count(cls):
        return cls.book_count

    @staticmethod
    def static_method_example():
        print("This is a static method.")
